microsoft/testsuites/dpdk/common.py

```diff
@@ -17,6 +22,261 @@
 AZ_ROUTE_ALL_TRAFFIC = "0.0.0.0/0"
 
 
+# Attempt to clean up the DPDK package dependency mess
+# Make a Installer class that implements the common steps
+# for installing DPDK/rdma-core, either from source or the package manager.
+# This generic class will get implemented in DpdkTestpmd and RdmaCore.
+# This should help us cover the various installation cases in a nice way,
+# and allow us to only re-implement the bits we need for each project.
+class OsPackageDependencies:
+    # A class to reduce the isinstance() trees that are
+    # sprinkled everywhere.
+    # Caller provides a function to match an OS and
+    # the packages to install on that OS.
+    def __init__(
+        self,
+        os_matcher: Callable[[Posix], bool],
+        packages: Sequence[Union[str, Tool, Type[Tool]]],
+    ) -> None:
+        self.check_node_os = os_matcher
+        self.packages = packages
+
+
+class DependencyInstaller:
+    # provide a list of OsPackageDependencies for a project
+    def __init__(self, requirements: List[OsPackageDependencies]) -> None:
+        self.requirements = requirements
+
+    # evaluate the list of package dependencies,
+    def install_required_packages(
+        self, node: Node, extra_args: Union[List[str], None]
+    ) -> None:
+        os = node.os
+        if not isinstance(os, Posix):
+            return
+        # find the first match for an OS, install the packages.
+        # Most specific should go first, most generic should go last.
+        for os_package_check in self.requirements:
+            if os_package_check.check_node_os(os):
+                os.install_packages(os_package_check.packages, extra_args=extra_args)
+                return
```

Comment:
It doesn't need to return, just check all of them. There maybe some reason to split packages in different ways.

Reply:
I'm not making more changes for a use case that isn't in-use for the PR. I have dependencies where I want to match for ubuntu 18.04 only, any other ubuntu, and then any other debian distro. It's cleaner to have these structured this way. If we ever need it the other way, we'll make that change later.

Reply:
We cannot change it later, once it's used by others. Because it's a breaking change on code behaviors.

Reply:
I mean more that if someone else is going to use this and they have a _requirement_ where they need the class to check every OS and not return, we can change it then. I don't see the need to change this _in anticipation_ of someone else using it.

Reply:
The dependency pattern is used in a couple of tools. Yes, we need to have a method to go through a list of requirements. The method is shared too.

Reply:
Ok, added a parameter to allow processing to continue

Reply:
follow up: already using this in a new PR 😂